### ISSUE TITLE
Monit closes Sprout when registering SNMP tables take too long

### ISSUE THREAD

**Issue by TygerBurning** [Metaswitch/sprout#1150]
When Sprout tries to register SNMP tables for statistics and the registration fails Monit resets Sprout. Registration fails during batch registration for multiple SNMP tables. There's some sort of choke when accessing OIDs in the relevant MIB file

```
29-07-2015 14:36:44.580 UTC Info snmp_table.h:87: Registering SNMP table sprout_homestead_mar_latency
29-07-2015 14:36:44.580 UTC Info snmp_table.h:87: Registering SNMP table sprout_homestead_sar_latency
29-07-2015 14:36:44.580 UTC Info snmp_table.h:87: Registering SNMP table sprout_homestead_uar_latency
29-07-2015 14:36:44.581 UTC Info snmp_table.h:87: Registering SNMP table sprout_homestead_lir_latency
29-07-2015 14:36:44.581 UTC Info snmp_table.h:87: Registering SNMP table initial_reg_success_fail_count
29-07-2015 14:36:44.581 UTC Info snmp_table.h:87: Registering SNMP table initial_reg_success_fail_count
29-07-2015 14:36:44.581 UTC Info snmp_table.h:87: Registering SNMP table initial_reg_success_fail_count
29-07-2015 14:36:45.602 UTC Error (Net-SNMP): registering pdu failed: 263!
29-07-2015 14:36:45.602 UTC Info snmp_table.h:87: Registering SNMP table initial_reg_success_fail_count
29-07-2015 14:36:46.621 UTC Info snmp_table.h:87: Registering SNMP table initial_reg_success_fail_count
29-07-2015 14:36:47.630 UTC Info snmp_table.h:87: Registering SNMP table initial_reg_success_fail_count
29-07-2015 14:36:48.650 UTC Info snmp_table.h:87: Registering SNMP table initial_reg_success_fail_count
29-07-2015 14:36:49.658 UTC Info snmp_table.h:87: Registering SNMP table initial_reg_success_fail_count
29-07-2015 14:36:50.666 UTC Info snmp_table.h:87: Registering SNMP table sprout_token_rate
```

```
Jul 29 14:39:23 ip-10-239-59-103 snmpd[9019]: duplicate registration: MIB modules AgentX subagent 84, session 0x2611a00, subsession 0x262a480 and AgentX subagent 84, session 0x2611a00, subsession 0x262a480 (oid .1.2.826.0.1.1578918.9.3.12).
Jul 29 14:39:33 ip-10-239-59-103 snmpd[9019]: duplicate registration: MIB modules AgentX subagent 84, session 0x2611a00, subsession 0x262a480 and AgentX subagent 84, session 0x2611a00, subsession 0x262a480 (oid .1.2.826.0.1.1578918.9.3.2.1).
```

**Comment by rkd-msw** [comment]
I don't see this when running the latest dev code:

```
29-07-2015 16:09:02.399 UTC Status snmp_agent.cpp:112: AgentX agent initialised
29-07-2015 16:09:02.400 UTC Info (Net-SNMP): NET-SNMP version 5.7.2 AgentX subagent connected
29-07-2015 16:09:02.400 UTC Info snmp_table.h:87: Registering SNMP table sprout_latency
29-07-2015 16:09:02.400 UTC Info snmp_table.h:87: Registering SNMP table sprout_queue_size
29-07-2015 16:09:02.400 UTC Info snmp_table.h:87: Registering SNMP table sprout_incoming_requests
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table sprout_rejected_overload
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table sprout_homestead_cxn_count
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table sprout_homestead_latency
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table sprout_homestead_mar_latency
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table sprout_homestead_sar_latency
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table sprout_homestead_uar_latency
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table sprout_homestead_lir_latency
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table initial_reg_success_fail_count
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table re_reg_success_fail_count
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table de_reg_success_fail_count
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table third_party_initial_reg_success_fail_count
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table third_party_re_reg_success_fail_count
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table third_party_de_reg_success_fail_count
29-07-2015 16:09:02.401 UTC Info snmp_table.h:87: Registering SNMP table sip_digest_auth_success_fail_count
29-07-2015 16:09:02.402 UTC Info snmp_table.h:87: Registering SNMP table ims_aka_auth_success_fail_count
29-07-2015 16:09:02.402 UTC Info snmp_table.h:87: Registering SNMP table non_register_auth_success_fail_count
29-07-2015 16:09:02.402 UTC Debug alarm.cpp:76: sprout cleared its alarms
```

Looks like https://github.com/Metaswitch/sprout/commit/a81886d248d253e0679a5d6c8d59987dfcacc900#diff-1a82dbcefa7b2c9ef254b0f885e726c2 fixed up the duplicate names - maybe those matter more than we thought?

**Comment by rkd-msw** [comment]
We've figured out that this is because we're calling `agent_check_and_process` too early - @alango is coding up a fix.

**Comment by eleanor-merry** [comment]
@alango Is this fixed?

**Comment by alango** [comment]
fixed by #1159
